internal/compiler-bridge/src/main/scala/xsbt/Dependency.scala

```diff
@@ -83,12 +105,11 @@ final class Dependency(val global: CallbackGlobal) extends LocateClassFile with
         def processDependency(context: DependencyContext)(dep: ClassDependency): Unit = {
           val fromClassName = className(dep.from)
           def binaryDependency(file: File, onBinaryClassName: String) =
-            callback.binaryDependency(file, onBinaryClassName, fromClassName, sourceFile, context)
+            withCallback(_.binaryDependency(file, onBinaryClassName, fromClassName, sourceFile, context))
```

Comment:
I would much rather see the thread-handling done inside sbt itself, so scalac synchronously calls `binaryDependency` but the implementation of `binaryDependency` itself will deal with threads